packages/fxa-react/lib/utils.tsx

```diff
@@ -12,3 +14,161 @@ export type FtlMsgProps = {
+        throw new Error(
+          `Could not locate message in en bundle. Message id: ${id}`
+        );
+      }
+
+      // Make sure the fallback text and message render cleanly.
+      checkMessage(enBundle, enMessage, fallbackText, args);
+    }
+
+    // If the above checks pass, we can safely return the language string.
+    return this.l10n.getString(id, args, fallbackText);
+  }
+}
+
+/**
+ * Validates a message. Checks that:
+ *  - fallback text matches ftl text
+ *  - there are not straight quotes
+ *  - there are no missing ftlArgs
+ * @param bundle - The default bundle, typically a bundle with 'en' locale.
+ * @param pattern - The pattern to check. This is the fluent messages's 'value'
+ * @param fallbackText - The fallback text
+ * @param ftlArgs - Arguments for updating parameterized messages
+ */
+export function checkMessage(
+  bundle: FluentBundle,
+  message: Message,
+  fallbackText: string | null,
+  ftlArgs?: Record<string, FluentVariable>
+) {
+  if (!fallbackText) {
+    throw new Error('Fallback text must be provided.');
+  }
+
+  // nested attributes can happen when we define something like:
+  // `profile-picture =
+  //   .header = Picture`
+  const nestedAttrValues = Object.values(message?.attributes || {});
+
+  if (
+    message === undefined ||
+    (message.value === null && nestedAttrValues.length === 0)
+  ) {
+    throw Error(`Invalid message. Message has no value.`);
+  }
+
+  if (message.value) {
+    _checkPattern(bundle, message.value, fallbackText, ftlArgs);
+  }
+
+  if (nestedAttrValues) {
+    nestedAttrValues.forEach((nestedAttrValue) =>
+      _checkPattern(bundle, nestedAttrValue, fallbackText, ftlArgs)
+    );
+  }
+}
+
+function _checkPattern(
+  bundle: FluentBundle,
+  pattern: Pattern,
+  fallbackText: string,
+  ftlArgs?: Record<string, FluentVariable>
+) {
+  const errors: Error[] = [];
+  const ftlMsg = bundle.formatPattern(pattern, ftlArgs, errors);
+
+  // Try to render the message, and check for any errors. This will catch things like missing parameters.
+  if (errors.length > 0) {
+    throw new Error(
+      `Errors encountered formatting fluent message. Fluent errors: ${errors
+        .map((x) => x.message)
+        .join('\n')}`
+    );
+  }
+
+  // We allow for .includes because fallback text comes from `textContent` within the `FtlMsg` wrapper which may contain
+  // more than one component and string. Also note that strings must be 'cleaned' because the ftlMsg may have invisible
+  // control characters that can break the includes check.
+  if (!_clean(fallbackText).includes(_clean(ftlMsg))) {
+    throw Error(
+      `Fallback text does not match Fluent message.\nFallback text: ${fallbackText}\nFluent message: ${ftlMsg}`
+    );
+  }
+
+  if (ftlMsg.includes("'")) {
+    throw Error(
+      `Fluent message contains a straight apostrophe (') and must be updated to its curly equivalent (’). Fluent message: ${ftlMsg}`
+    );
+  }
+
+  if (ftlMsg.includes('"')) {
```

Comment:
Q: how would this handle .ftl strings w/ markup, like:

```ini
fxa-header-firefox-logo = <img data-l10n-name="fxa-logo" alt="{ -brand-firefox } logo">
fxa-header-sync-devices-image = <img data-l10n-name="sync-devices-image" alt="Sync devices">
body-devices-image = <img data-l10n-name="devices-image" alt="Devices">
```

https://github.com/mozilla/fxa-content-server-l10n/blob/main/locale/en/auth.ftl#L48-L50

---

Didn't see any obvious uses of straight single quotes in fxa-content-server-l10n en locale [that _weren't_ comments] using this grep:
```sh
git grep -In "'" locale/en/**/*.ftl | grep -v ":\#"
```

Reply:
Hi Peter, thanks for pointing out this case.  @LZoog is this the `DOM Overlay` case you were talking about in the ticket? If so, and we want to support this, I think straight quotes would be acceptable here. I can add test coverage for this, although it does make the parsing a bit more complicated.

Reply:
I haven't had a chance to check out the PR, but yes this is what I meant by the DOM Overlay test case. I didn't realize in my previous test set up PR we'd need this case until I'd basically already finished it. 🙈

Reply:
@dschom 🙌 I'm definitely going to be taking a closer look at all of this tomorrow but just wanted to leave a fly-by comment after looking at #14442 (I reference your PR in a comment there, I think when Valerie rebases she'll need to change one line) and say, **this PR is awesome**. 🎉 Thanks for all of the effort.